File: neetcode-250/sliding-window/121-best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
	let maxProfit = 0,
		left = 0;

	for (let right = 1; right < prices.length; ++right) {
		const buyPrice = prices[left],
			sellPrice = prices[right];

		if (buyPrice > sellPrice) {
			left = right;
		} else {
			maxProfit = Math.max(maxProfit, sellPrice - buyPrice);
		}
	}

	return maxProfit;
}

File: neetcode-250/sliding-window/121-best-time-to-buy-and-sell-stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res, l = 0, 0
        for r in range(1, len(prices)):
            buy_price, sell_price = prices[l], prices[r]
            if sell_price < buy_price:
                l = r
            else:
                res = max(res, sell_price - buy_price)
        return res
